### ISSUE TITLE
Re-engineer `main`

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#3338]
We have discussed this on-and-off in person and online.

The proposal is: lets give `main` a much simpler type signature such that programmers can easily get started with Flix. 

Update `main` to simply be written as:

```scala
def main(): Unit = println(...)
```

~which can even be shorted to:~

```scala
def main() = println(...)
```

To access the arguments array you simply call `Environment.getArguments()` and you get the array.

We need the following changes:

- [ ] ~Add `Environment.getArguments()` which calls `System.getProperty("args")`.~
- [ ] ~Modify `GenMain` to call `System.setProperty("args", args)` to store the arguments array.~
- [ ] If `def main() = exp1` then rewrite that to `def main() = exp1; ()`.
- [ ] Update the type checking and inference for `main`.
- [ ] Change auto-complete snippet in LSP.
- [ ] Add a check that `main` cannot be called from other functions.
- [ ] ~Add a function `System.exit` which calls the Java equivalent.~

(EDIT: This is partially inspired by #3327 which reveals that we cannot really have a exit code on the JVM).

**Comment by magnus-madsen** [comment]
@mlutze @JonathanStarup @jaschdoc @ninaandrup @stephentetley Any objections or comments?

**Comment by mlutze** [comment]
I think we should make the property `flix.args` or something like that so there's a smaller chance of collision.

**Comment by jaschdoc** [comment]
100% agree with this. Would it make sense to have a global `SysOptions` enum and then call `System.getProperty(SysOpts.args)` or something like that to avoid having magic strings? Haven't thought this through completely

**Comment by magnus-madsen** [comment]
The string would never be seen by the programmer. Its internal to the Flix compiler / library.

**Comment by jaschdoc** [comment]
Oh right, just re-read the post. :smile:

**Comment by magnus-madsen** [comment]
@JonathanStarup If you could do the change to the backend first then the rest of us can do the various other things.

**Comment by JonathanStarup** [comment]
1) I am opposed to adding system exit to the flix lib since it easy enough to import and I don't think we should encourage use
2) what should I change on the backend? Not call sys exit?

**Comment by ninaandrup** [comment]
No objections. I think it looks very nice and simple! :)

**Comment by magnus-madsen** [comment]
>     2. what should I change on the backend? Not call sys exit?

Our main (which calls the programs main) need to store the argument array by calling `System.setProperty("args", args)` so we can retrieve it later.

**Comment by stephentetley** [comment]
Looks good!

At some point I'd like to write a successor to GetOpt as I find it difficult to use, but if the args are accessible as an array that's fine for GetOpt (or its successor).

**Comment by magnus-madsen** [comment]
Further proposal:

- Add a command line argument (and LSP argument) `--main foo` which specifies the name of main.
- If the argument is absent it defaults to `main`.
- If the argument is present the `Namer` (`Resolver`?) will check that it exists. 
- When compiling we always use the given `main` as the entry point.

The upshot is that we can run multiple functions "as main" in VSCode.

**Comment by JonathanStarup** [comment]
Recent changes to ideas:
* At some point, check that the main-named function is never called
* Check that the main-named function has 1 argument of either Unit or Array[String] and change Unit arg to Array[String] arg named _
* If the return type is omitted, change it to Unit and insert `...; ()` to make the returned value fit

**Comment by magnus-madsen** [comment]
New list of tasks:

- [ ] Rewrite `def main()` to `def main(_: Array[String])`
- [ ] Change the expected signature of `main` to be `Unit & _` where `_` means any effect.
- [ ] Wrap the body of `main` in a `exp; ()` expression statement. 

@mlutze Which phases do you think this belongs in?

**Comment by mlutze** [comment]
I am not sure we can reliably identify main before Resolver, so transformations would probably go there.

We have an existing `IllegalMain` check in the Typer, but I think it's okay to keep them separate. It has to be in the Typer so that the type of main can be checked.

We'll also need to modify the Parser to handle missing return type.

**Comment by magnus-madsen** [comment]
> We'll also need to modify the Parser to handle missing return type.

I suggest we skip that for now. It seems more irregular than the rest.

**Comment by magnus-madsen** [comment]
@mlutze Do you want to take a stab at this? I think the backend support should be there now. 

So just to recap, the idea is that you can write `def main` with or without the array argument. Further, main should now return `Unit` (which we ensure by wrapping it in an expression statement).

**Comment by magnus-madsen** [comment]
Its not urgent.

**Comment by JonathanStarup** [comment]
I think it would be a lot nicer to use this if it is also supported inside namespaces. I was thinking about a hypothetical project with five namespaces that I want to have an entry point.
* In the current version I have to create a single top-level `def main` that uses args to programmatically find the entry point
* In this version each namespace have to make a uniquely named main function in the top-level. If the five namespaces are really their own little script/gen thing I'd like to call all of them main but I can't. This means that I'd probably name them `def mainNamespace`. I don't like this since it creates top-level functions that should never be called and gives compilation errors dependent on flags.
* In relation to these functions that float around but cannot be called, I think we should add an optional `@entrypoint` annotation that asserts to the compiler that calling this function is a compilation error. This allows programmers to avoid introducing errors that only appear on some main flags. This could possibly also be used in vscode to remove them from autocomplete.

In conclusion: I think we should allow the custom main to be inside namespaces and add a `@entrypoint` annotation that disallows calls to the function.

**Comment by magnus-madsen** [comment]
I think we can allow the entrypoint to be any `Name.QName / DefnSym`.

**Comment by magnus-madsen** [comment]
Tasks:
- Magnus: I need to allow all ASTs to carry a named entry point.
- Matt: Will allow main to be more lenient with its types.
- Jonathan: Will help with the backend, if anything is needed.

Finally, I think we need @nicdard to help with the VSCode integration (but first when all of the above is done).

**Comment by nicdard** [comment]
Ok!

**Comment by magnus-madsen** [comment]
@nicdard We can work a bit in parallel. Once https://github.com/flix/flix/pull/3430 is merged, we know have:

![image](https://user-images.githubusercontent.com/2287953/160580121-f6ee4202-1fa6-4646-afa4-a5abf6d05807.png)

What we need, and I hope you can help with, is some new command that we can invoke which stores the name of the specific function you are clicking on. Can we preserve the name in the command https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/api/lsp/Command.scala#L29 ?

Basically when I click on "Run TODO" we need to pass e.g. "baz" to the command that invokes the Flix compiler.

**Comment by magnus-madsen** [comment]
@nicdard The extra information is attached here: https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/api/lsp/provider/CodeLensProvider.scala#L48 I wonder how you can extract it on the VSCode side.

**Comment by magnus-madsen** [comment]
In #3433 I have added a `--entrypoint <name>` flag. If this flag is set then `name` is used as the main entry point. What we need now is two-fold: (A) A VSCode command that when you click on a CodeLenses passes this argument as `--entrypoint` and (B) relaxing the type requirements (and redundancy?) of main.

**Comment by magnus-madsen** [comment]
I think all we need to do is to add an `args` argument here:

https://github.com/flix/vscode-flix/blob/master/client/src/handlers/handlers.ts#L207

and then ensure we build a command with `... --entrypoint args`.

**Comment by magnus-madsen** [comment]
A JAR to test with is available here: https://github.com/flix/flix/pull/3433

**Comment by nicdard** [comment]
@magnus-madsen just to recap, sorry for writing back just now, long day at uni, you have already implemented the argument passing that you were asking today right? I there anything left to do for which I am of any help? :)

**Comment by magnus-madsen** [comment]
Yep! I was having fun :)

For this ticket, I don't think so.

**Comment by nicdard** [comment]
ahah ok! :)

**Comment by magnus-madsen** [comment]
@mlutze Do you want to take a stab at being more permission on the signature of main?

I think we want three things (not sure how difficult each is):
- Allow the programmer to omit the args array.
- Allow the programmer to omit the effect.
- Allow the programmer to return any type.

(Possibly the last two can be solved by wrapping the main expression-- not sure if this has unintended consequences).

**Comment by magnus-madsen** [comment]
What if you have:

```
def myTest01(): Bool = "hello" == "bar"
```

We will now be able to click to run it, but we wont see the result. Thoughts?

**Comment by mlutze** [comment]
Could we magically check if the result is an instance of `ToString` and use that?

**Comment by magnus-madsen** [comment]
Possible? And if not?

Alternatively:

- If return type is `Unit` we write to `exp; ()`. 
- If return type is `non-Unit` we write to `println(exp); ()`

Might be simpler and easier to understand.

Then `def main(): MyType` would ask for a `ToString` instance which seems reasonable.

**Comment by JonathanStarup** [comment]
If we print the output then main also do not have to be pure I suppose

**Comment by magnus-madsen** [comment]
I think we should do this:

- If return type is `Unit` we write to `exp; ()`. 
- If return type is `non-Unit` we write to `println(exp); ()`

@mlutze Do you want to implement it?

**Comment by magnus-madsen** [comment]
While also allowing main to take Unit or Array[String] as arguments.

**Comment by mlutze** [comment]
I'm finding that it's rather complex to modify the code in-place, as we have to juggle type schemes, etc. while maintaining source locations wherever possible... and I think it will end up being very brittle.

I'm thinking as an alternative, we can generate a main function that wraps the entry point. This way the locations and types of the entrypoint stay the same, and there's not as much disconnect from the source code.

**Comment by magnus-madsen** [comment]
Let's discuss more offline . Did you try it in the weeder or much later?

**Comment by mlutze** [comment]
I tried in the Resolver and then the Typer.

I don't think Weeder is a good idea, since we likely end up with more "early desugaring" problems. E.g. a user could end up with something like "cannot generalize this to `Array[String] -> Unit`" even though the signature says `Unit -> Unit`.

**Comment by JonathanStarup** [comment]
will you let me know if I should move args to a global class @magnus-madsen?

**Comment by magnus-madsen** [comment]
> will you let me know if I should move args to a global class @magnus-madsen?

Yes, but for now I don't think we need it.

**Comment by magnus-madsen** [comment]
I have decided that I want to store the args array and add a `Environment.getArguments()`.

**Comment by magnus-madsen** [comment]
@JonathanStarup Can I ask you to add support for the following (inspired by GlobalCounter):

Add a class `GlobalEnvironment` which is approximately:

```
public final class GlobalEnvironment {
    private static final String[] args;

    public static final String[] getArgs() {
        return args;
    }
}
```

Add a `GenGlobalEnvironment` class. 

Modify `GenMain` to store the args into the GlobalEnvironment class.

Thanks!

**Comment by mlutze** [comment]
System.arrayCopy of args instead of the actual args?

**Comment by magnus-madsen** [comment]
> System.arrayCopy of args instead of the actual args?

Lets do that in Flix code, because its a lot of work to generate the bytecode for it. Also, obv. the args field cannot be final.

**Comment by magnus-madsen** [comment]
```
@Internal
pub def printResult(x: a): Unit & Impure with ToString[a] = 
    match reifyType a {
        case ReifiedInt8 => ()
        case _           => println(x)
    }
```

**Comment by magnus-madsen** [comment]
- Explore moving this phase to before typing
- Refactor GetOpt to use list rather than array
- Update entrypoint-related stuff in LSP
- Change existing Int32 return types to Unit

**Comment by magnus-madsen** [comment]
@JonathanStarup I have merged Matt's changed. Will you cleanup the backend? Just remove everything that is no longer needed.

**Comment by magnus-madsen** [comment]
Are we done or?

**Comment by mlutze** [comment]
I had one more idea to complicate things: an `ExitCode` type (just wrapping `Int`) so that if we want to be sure an explicit exit code is given, we can enforce it with the type.

**Comment by magnus-madsen** [comment]
Dont understand

**Comment by mlutze** [comment]
```scala
def main(): ExitCode =
    if(everythingOk())
        ExitCode(0)
    else
        ExitCode(1)
```

So instead of calling Exit in each branch (and maybe forgetting a branch so we implicitly get success) we return an ExitCode which is used as the program exit code.

**Comment by magnus-madsen** [comment]
But that would make things more complicated right?

**Comment by mlutze** [comment]
Yes, though if we get default instances it's a lot easier. Either way we can leave it for future work.
